sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -99,6 +99,17 @@ object TypeCoercion {
     case (_: TimestampType, _: DateType) | (_: DateType, _: TimestampType) =>
       Some(TimestampType)
 
+    case (t1 @ ArrayType(pointType1, nullable1), t2 @ ArrayType(pointType2, nullable2))
+        if t1.sameType(t2) =>
+      val dataType = findTightestCommonType(pointType1, pointType2).get
```

Comment:
If `t1` and `t2` are the sameType, why do we need recursively `findTightestCommonType`? Can we just make `val dataType = pointType1`?

Reply:
I think that the reason is for nested structures, for instance an array of array or array of map

Reply:
then @bdrillard can we add also some test cases for nested structures?

Reply:
Let's use this example, if we call `findTightestCommonType` on 2 array of array of Integer types, then this `dataType` = `findTightest(findTightest(IntegralType, IntegralType))` = `IntegralType`, so the final answer is `Some(ArrayType(IntegralType))`?

Reply:
@mgaido91, @gczsjdy, sure, I'll add a pair of test cases for nested complex structures for `ArrayType` and `MapType`

Reply:
@mgaido91 Thanks, I got it.

Reply:
@gczsjdy  I think it is needed, because IIUC in `sameType` the nullability is not checked to be the same. So it might vary and calling `findTightestCommonType` does the trick.

Reply:
@gczsjdy `sameType` will ignore the nullability of the two types, only checking if the `DataType` is the same. We would expect then that the first case of `findTightestCommonType` would pass the types on through, but during its equality check, it also checks the type nullability, which for ArrayType and MapType, may differ between t1 and t2:

```
case (t1, t2) if t1 == t1 => Some(t1)
```

That means we can establish that two StructFields of ArrayType/MapType are the same DataType, but if they have different nullability, then the above case match won't match them, nor will any other case in the match set. So in order to find the tightest common type where nullabilities of the point-types may differ, we'll need to recurse. See a case like:

```
widenTest(
    StructType(StructField("a", ArrayType(StringType, containsNull=true)) :: Nil),
    StructType(StructField("a", ArrayType(StringType, containsNull=false)) :: Nil),
    Some(StructType(StructField("a", ArrayType(StringType, containsNull=true)) :: Nil)))
```

At the moment, since we have no case to match ArrayType/MapType where the nullabilities may differ, this case would fail. I can add this test explicitly to the suite.

Reply:
Thanks for your illustration.